Refactor to introduce suggested better practice across React and Ably

leggetter opened the pull request:
Main changes:

- Publish takes place via a `fetch` to `/api/submit-news` (maybe it should be just `/news`) which means some validation/verification of the content could take place on the server vs. publishing on the client. If we agree with this change then we should use a subscribe only key on the client.
- Componentized as much as possible

Note: I may have broken something related to showing the `author` within the news headline. I'm not sure how this worked.

Additional consideration: would submitting a URL and then using Open Graph ([Node.js package](https://github.com/samholmes/node-open-graph)) to get a preview title reflect a more realistic and interesting use case?

leggetter committed:
fix: remove duplicate ABLY_API_KEY from example.env

leggetter committed:
refactor: remove "Component" filaname suffix

leggetter committed:
refactor: remove unused "name" from .publish

leggetter committed:
feat: submit news to server to publish

leggetter committed:
refactor: move date formatting into function

leggetter committed:
refactor: convert headlinePreviews function into component

leggetter committed:
chore: remove console.log

marklewin reviewed (commented):
Hi @leggetter . Thanks for the PR, but I don't think moving the publishing to the server will work for this example (and, in fact, it doesn't seem to: the headlines aren't being published to the `news-list` channel). Your endpoint (`api/submit-news`) is instantiating the REST SDK without a valid `clientId`, which is required for this channel in order to specify who created what. Also, I can't help but think that we're complicating things, by having two different SDKs at work here. WDYT?

leggetter commented:
Hey @marklewin,

> Your endpoint (api/submit-news) is instantiating the REST SDK without a valid clientId, which is required for this channel in order to specify who created what.

Ah, that addresses my note of:

> Note: I may have broken something related to showing the author within the news headline. I'm not sure how this worked.

I'd suggest we update to either:

1. Send the client ID of the publisher to the server and use it there (I don't know if we support this)
2. Make `author` a part of the data payload. Since the author is data this feels like the right approach.

To address some questions:

> I can't help but think that we're complicating things, by having two different SDKs at work here. WDYT?

IMO only slightly. A few key points:

1. I know it's only a demo but with a little bit more complexity we're showing a better practice (don't trust your client and in most cases you should validate and publish via the server)
3. Showing how to publish which I believe will be most common with Next.js (publishing via an /api route).

We are only using one SDK. We're using it once directly (server) and the second time we use it via the Ably React Hooks package.

leggetter committed:
Move author into data payload

- passing author to the server
- removed passing HeadlineText since it's a global anyway
- update server to us publishing single event

leggetter commented:
@marklewin - I've pushed a commit that also sends an `author` property in the payload to the server that's also used with the published data.

Let me know your thoughts on this and the query about complexity.

leggetter commented:
Additional thought: It may be possible to introduce the following later in the blog post as a way of layering on best practice:

1. Publishing via the server
2. Token auth instead of API Key

Maybe we're at a point where we should move to creating an outline for the post?

I do think the OpenGraph idea adds a bit of richness to the demo though so adding a `/api` endpoint for that and asking for a URL from the user instead of headline is worth considering.

marklewin commented:
Thanks @leggetter for the comments and new commit. I've thought about everything you suggest here and think it will make for a better tutorial if we can implement your suggestions. I'll attempt to achieve this today, with the aim of working on the post outline tomorrow. Sound good?

leggetter commented:
@marklewin sounds great. Thank you!

marklewin commented:
@leggetter I’ve dug into this a little bit this morning and, in the meantime, refreshed my knowledge about how Ably authentication works :)

When I built this, I was under the impression (after a chat with @thisisjofrank ) that we should be advocating for the use of Ably React Hooks generally, for both publishing and subscribing. I believe that your idea of moving the publishing side of things to the server like you have contradicts the way that the Ably React Hooks package wants to manage things.

If you run your PR, you’ll see that it doesn’t work. The presence list is populated correctly, but the headlines aren’t being published. That’s because the REST client you are instantiating on the server (just for publishing) hasn’t been created with the same `clientId` that is used in the React Hooks’ `configureAbly()` method (or any `clientId`). Without an identified client, the channel won’t let the server instance publish to it.

In order for the headlines to show the correct author, the server would have to use the same `clientId` that was used by ‘configureAbly()`. It’s not enough to just create the REST instance with a client ID, you would have to use the `rest.auth.createTokenRequest({clientId: “Alice”}, …` method to generate its own identifying token.

I’m sure there’s a hacky way to get around this if we had to, but it would make for quite a confusing tutorial IMO.

I think a better approach would be to continue to use Ably React Hooks to do the publishing, but to manage the authentication and generation of client IDs and tokens via an `authUrl`. I believe that would be a much nicer implementation and still show good practice (issuing token requests and client Ids from the server, rather than from the client).

WDYT?

marklewin commented:
@leggetter I have created a [new PR](https://github.com/ably-labs/ably-next-vercel-news/pull/3) to implement the changes I propose above (token reqs and clientId issued by the API). Please let me know your thoughts. If it looks OK, I can try and implement your OpenGraph suggestion?

leggetter commented:
> When I built this, I was under the impression (after a chat with @thisisjofrank ) that we should be advocating for the use of Ably React Hooks generally, for both publishing and subscribing. I believe that your idea of moving the publishing side of things to the server like you have contradicts the way that the Ably React Hooks package wants to manage things.

I don't believe it does contradict the way the Ably React Hooks package wants to manage things. We're just not using it for the publishing part because we want to send the data to the /api route first to be validated and to show how you publish in that way. 

> If you run your PR, you’ll see that it doesn’t work. The presence list is populated correctly, but the headlines aren’t being published. That’s because the REST client you are instantiating on the server (just for publishing) hasn’t been created with the same clientId that is used in the React Hooks’ configureAbly() method (or any clientId). Without an identified client, the channel won’t let the server instance publish to it.

In https://github.com/ably-labs/ably-next-vercel-news/pull/2/commits/2e78a47846d914d334218e386e7c4e0fc09c7cf5 I added the `author` to the payload which removes the need to know the Client ID when publishing.

I've run this locally and "it works for me". Did you pull this recent commit?

<img width="1511" alt="image" src="https://user-images.githubusercontent.com/328367/165800030-b74b79e0-cc4b-46f6-b37c-6bf51239142f.png">

> I think a better approach would be to continue to use Ably React Hooks to do the publishing, but to manage the authentication and generation of client IDs and tokens via an authUrl. I believe that would be a much nicer implementation and still show good practice (issuing token requests and client Ids from the server, rather than from the client).

I disagree about the approach to publishing for the reasons stated. So maybe we do need to get another opinion 😊 

I agree on the introduction of the use of tokens.

Maybe we can jump on a call with the additional set of 👀, talk it through, and come to a decision.

marklewin commented:
Thanks @leggetter . This [video](https://www.loom.com/share/0721201a6822470c8f177e2812cbaf2a) shows what's happening for me.

I'm very open to the idea I might be wrong 😊 . So, I agree it would be useful to jump on a call with @thisisjofrank ... I'll ask if she can spare some time to talk tomorrow!

leggetter commented:
@marklewin The video shows nothing is working 🤔  I've definitely pushed everything to the branch.

Have a look at your DevTools network tab to see if the POST request to `/api/submit-news` is working.

<img width="868" alt="image" src="https://user-images.githubusercontent.com/328367/165811288-2f18c732-bef9-4d87-89dc-fb88aadc10e4.png">

You could also add some logging to the handler to see what's happening at the point of publishing.

I'm pulling my config from a `.env.local` rather than a `.env`. I doubt that's it but it's something different between the setups.

marklewin commented:
Thanks @leggetter, but the issue appears to be that I have a channel rule set up on `news-list` to require [identified clients](https://ably.com/docs/core-features/authentication#identified-clients). I'm guessing that you don't, because when I remove the rule, your app works 😄  Let's chat about it in our meeting this morning and I'll report back here.

thisisjofrank commented:
Do you still need my input here?

* Doing some work on the server does add to the vercel focused nature of the demo
* There is no problem with putting the publisher in the data, this seems sensible
* It only makes sense to do the work on the server if we do the opengraph title abstraction, because otherwise it feels functionally identical to publishing from the client.
* For the sake of this demo it might be easier to fetch the page rather than introducing an additional dependency (for the sake of keeping the demo and tutorial simple and dependency list shorter)

marklewin commented:
Closing, due to these changes being incorporated in https://github.com/ably-labs/ably-next-vercel-news/pull/7

leggetter commented:
Mobile at the mo. But have you pulled the latest commit? Just checking.

On Thu, 28 Apr 2022, 12:10 Mark Lewin, ***@***.***> wrote:

> @leggetter <https://github.com/leggetter> I’ve dug into this a little bit
> this morning and, in the meantime, refreshed my knowledge about how Ably
> authentication works :)
>
> When I built this, I was under the impression (after a chat with
> @thisisjofrank <https://github.com/thisisjofrank> ) that we should be
> advocating for the use of Ably React Hooks generally, for both publishing
> and subscribing. I believe that your idea of moving the publishing side of
> things to the server like you have contradicts the way that the Ably React
> Hooks package wants to manage things.
>
> If you run your PR, you’ll see that it doesn’t work. The presence list is
> populated correctly, but the headlines aren’t being published. That’s
> because the REST client you are instantiating on the server (just for
> publishing) hasn’t been created with the same clientId that is used in
> the React Hooks’ configureAbly() method (or any clientId). Without an
> identified client, the channel won’t let the server instance publish to it.
>
> In order for the headlines to show the correct author, the server would
> have to use the same clientId that was used by ‘configureAbly(). It’s not
> enough to just create the REST instance with a client ID, you would have to
> use the rest.auth.createTokenRequest({clientId: “Alice”}, …` method to
> generate its own identifying token.
>
> I’m sure there’s a hacky way to get around this if we had to, but it would
> make for quite a confusing tutorial IMO.
>
> I think a better approach would be to continue to use Ably React Hooks to
> do the publishing, but to manage the authentication and generation of
> client IDs and tokens via an authUrl. I believe that would be a much
> nicer implementation and still show good practice (issuing token requests
> and client Ids from the server, rather than from the client).
>
> WDYT?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/ably-labs/ably-next-vercel-news/pull/2#issuecomment-1112076641>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AACQFL4MS525ITH43GGTCSTVHJW4BANCNFSM5UKVSPOA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

marklewin commented:
Yes, this was closed and dealt with in https://github.com/ably-labs/ably-next-vercel-news/pull/7
